File: Android/app/src/main/java/com/auralis/crisisconnect/screens/settings/ProfileCertificateCard.kt
package com.auralis.crisisconnect.screens.settings

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AccessTime
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.ErrorOutline
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.WarningAmber
import androidx.compose.material.icons.outlined.VerifiedUser
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.auralis.crisisconnect.R
import java.text.DateFormat
import java.util.Date
import java.util.Locale

/**
 * Renders the device-bound rescue certificate. State and actions live on
 * [ProfileViewModel] so the profile screen header (role chip, verified
 * badge) stays in sync with revoke / provision actions taken here.
 */
@Composable
fun ProfileCertificateCard(
    viewModel: ProfileViewModel,
    modifier: Modifier = Modifier,
) {
    val state by viewModel.certificateState.collectAsStateWithLifecycle()
    var confirmRevoke by remember { mutableStateOf(false) }

    LaunchedEffect(viewModel) { viewModel.refreshCertificate() }

    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = MaterialTheme.shapes.large,
        color = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.45f),
        contentColor = MaterialTheme.colorScheme.onSurfaceVariant,
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outline.copy(alpha = 0.25f)),
    ) {
        Column(modifier = Modifier.padding(horizontal = 18.dp, vertical = 18.dp)) {
            CardHeader(state = state)

            Spacer(modifier = Modifier.height(14.dp))

            when (val status = state.status) {
                is CertificateStatus.Loading -> LoadingBlock()
                is CertificateStatus.Missing -> MissingBlock(
                    provisioning = state.provisioning,
                    onProvision = { viewModel.provisionCertificate() },
                )
                is CertificateStatus.Loaded -> LoadedBlock(
                    status = status,
                    revoking = state.revoking,
                    provisioning = state.provisioning,
                    onRefresh = { viewModel.refreshCertificate() },
                    onRevokeClicked = { confirmRevoke = true },
                    onRenew = { viewModel.provisionCertificate() },
                )
                is CertificateStatus.Failure -> FailureBlock(
                    message = status.message,
                    onRetry = { viewModel.refreshCertificate() },
                )
            }

            state.statusMessage?.let { message ->
                Spacer(modifier = Modifier.height(10.dp))
                InfoBanner(
                    text = message,
                    isError = false,
                    onDismiss = { viewModel.consumeCertificateMessages() },
                )
            }
            state.errorMessage?.let { message ->
                Spacer(modifier = Modifier.height(10.dp))
                InfoBanner(
                    text = message,
                    isError = true,
                    onDismiss = { viewModel.consumeCertificateMessages() },
                )
            }
        }
    }

    if (confirmRevoke) {
        AlertDialog(
            onDismissRequest = { if (!state.revoking) confirmRevoke = false },
            icon = {
                Icon(
                    imageVector = Icons.Filled.WarningAmber,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.error,
                )
            },
            title = { Text(text = stringResourceSafe(R.string.profile_cert_revoke_title)) },
            text = { Text(text = stringResourceSafe(R.string.profile_cert_revoke_body)) },
            confirmButton = {
                TextButton(
                    enabled = !state.revoking,
                    onClick = {
                        confirmRevoke = false
                        viewModel.revokeCertificate(reason = "user_request")
                    },
                ) {
                    Text(
                        text = stringResourceSafe(R.string.profile_cert_revoke_confirm),
                        color = MaterialTheme.colorScheme.error,
                    )
                }
            },
            dismissButton = {
                TextButton(
                    enabled = !state.revoking,
                    onClick = { confirmRevoke = false },
                ) {
                    Text(text = stringResourceSafe(R.string.profile_cert_cancel))
                }
            },
        )
    }
}

@Composable
private fun CardHeader(state: CertificateUiState) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Surface(
            shape = CircleShape,
            color = MaterialTheme.colorScheme.primary.copy(alpha = 0.18f),
            contentColor = MaterialTheme.colorScheme.primary,
        ) {
            Icon(
                imageVector = Icons.Outlined.VerifiedUser,
                contentDescription = null,
                modifier = Modifier
                    .padding(10.dp)
                    .size(20.dp),
            )
        }
        Spacer(modifier = Modifier.width(12.dp))
        Column(modifier = Modifier.fillMaxWidth().weight(1f, fill = false)) {
            Text(
                text = stringResourceSafe(R.string.profile_cert_card_title),
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold,
            )
            Text(
                text = stringResourceSafe(R.string.profile_cert_card_subtitle),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.75f),
            )
        }
        val status = state.status
        if (status is CertificateStatus.Loaded) {
            StatusChip(status = status)
        }
    }
}

@Composable
private fun StatusChip(status: CertificateStatus.Loaded) {
    val (label, container, content) = when {
        status.isRevoked -> Triple(
            stringResourceSafe(R.string.profile_cert_status_revoked),
            MaterialTheme.colorScheme.errorContainer,
            MaterialTheme.colorScheme.onErrorContainer,
        )
        status.isExpired -> Triple(
            stringResourceSafe(R.string.profile_cert_status_expired),
            MaterialTheme.colorScheme.errorContainer,
            MaterialTheme.colorScheme.onErrorContainer,
        )
        else -> Triple(
            stringResourceSafe(R.string.profile_cert_status_active),
            MaterialTheme.colorScheme.primaryContainer,
            MaterialTheme.colorScheme.onPrimaryContainer,
        )
    }
    Surface(
        shape = MaterialTheme.shapes.small,
        color = container,
        contentColor = content,
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(horizontal = 10.dp, vertical = 4.dp),
        ) {
            val icon = when {
                status.isRevoked -> Icons.Filled.Close
                status.isExpired -> Icons.Filled.AccessTime
                else -> Icons.Filled.CheckCircle
            }
            Icon(imageVector = icon, contentDescription = null, modifier = Modifier.size(14.dp))
            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = label,
                style = MaterialTheme.typography.labelMedium,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}

@Composable
private fun LoadingBlock() {
    Row(verticalAlignment = Alignment.CenterVertically) {
        CircularProgressIndicator(modifier = Modifier.size(16.dp), strokeWidth = 2.dp)
        Spacer(modifier = Modifier.width(10.dp))
        Text(
            text = stringResourceSafe(R.string.profile_cert_loading),
            style = MaterialTheme.typography.bodyMedium,
        )
    }
}

@Composable
private fun MissingBlock(provisioning: Boolean, onProvision: () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Text(
            text = stringResourceSafe(R.string.profile_cert_missing_body),
            style = MaterialTheme.typography.bodyMedium,
        )
        Button(
            onClick = onProvision,
            enabled = !provisioning,
            modifier = Modifier.fillMaxWidth(),
            shape = MaterialTheme.shapes.large,
        ) {
            if (provisioning) {
                CircularProgressIndicator(
                    modifier = Modifier.size(16.dp),
                    strokeWidth = 2.dp,
                    color = MaterialTheme.colorScheme.onPrimary,
                )
                Spacer(modifier = Modifier.width(10.dp))
                Text(text = stringResourceSafe(R.string.profile_cert_request_in_progress))
            } else {
                Icon(imageVector = Icons.Filled.Lock, contentDescription = null)
                Spacer(modifier = Modifier.width(8.dp))
                Text(text = stringResourceSafe(R.string.profile_cert_request_button))
            }
        }
    }
}

@Composable
private fun LoadedBlock(
    status: CertificateStatus.Loaded,
    revoking: Boolean,
    provisioning: Boolean,
    onRefresh: () -> Unit,
    onRevokeClicked: () -> Unit,
    onRenew: () -> Unit,
) {
    val df = remember {
        DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.SHORT, Locale.getDefault())
    }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        DetailRow(
            label = stringResourceSafe(R.string.profile_cert_role_label),
            value = status.certificate.role,
        )
        DetailRow(
            label = stringResourceSafe(R.string.profile_cert_device_label),
            value = status.certificate.deviceId,
            mono = true,
        )
        DetailRow(
            label = stringResourceSafe(R.string.profile_cert_issued_label),
            value = df.format(Date(status.certificate.issuedAtMillis)),
        )
        DetailRow(
            label = stringResourceSafe(R.string.profile_cert_expires_label),
            value = df.format(Date(status.certificate.expiresAtMillis)),
            highlight = when {
                status.isExpired -> MaterialTheme.colorScheme.error
                status.certificate.expiresAtMillis - System.currentTimeMillis() <
                    24L * 60 * 60 * 1000 -> MaterialTheme.colorScheme.tertiary
                else -> null
            },
        )

        HorizontalDivider(
            modifier = Modifier.padding(vertical = 4.dp),
            color = MaterialTheme.colorScheme.outline.copy(alpha = 0.15f),
        )

        val canRenew = status.isExpired
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            if (canRenew) {
                Button(
                    onClick = onRenew,
                    modifier = Modifier.weight(1f),
                    enabled = !provisioning,
                    shape = MaterialTheme.shapes.large,
                ) {
                    if (provisioning) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(16.dp),
                            strokeWidth = 2.dp,
                            color = MaterialTheme.colorScheme.onPrimary,
                        )
                    } else {
                        Icon(imageVector = Icons.Filled.Refresh, contentDescription = null)
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(text = stringResourceSafe(R.string.profile_cert_renew_button))
                    }
                }
            } else {
                OutlinedButton(
                    onClick = onRefresh,
                    modifier = Modifier.weight(1f),
                    shape = MaterialTheme.shapes.large,
                ) {
                    Icon(imageVector = Icons.Filled.Refresh, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(text = stringResourceSafe(R.string.profile_cert_refresh_button))
                }
            }
            OutlinedButton(
                onClick = onRevokeClicked,
                modifier = Modifier.weight(1f),
                enabled = !revoking && !status.isRevoked,
                border = BorderStroke(
                    1.dp,
                    MaterialTheme.colorScheme.error.copy(alpha = 0.5f),
                ),
                colors = ButtonDefaults.outlinedButtonColors(
                    contentColor = MaterialTheme.colorScheme.error,
                ),
                shape = MaterialTheme.shapes.large,
            ) {
                if (revoking) {
                    CircularProgressIndicator(
                        modifier = Modifier.size(16.dp),
                        strokeWidth = 2.dp,
                        color = MaterialTheme.colorScheme.error,
                    )
                } else {
                    Icon(imageVector = Icons.Filled.Close, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(text = stringResourceSafe(R.string.profile_cert_revoke_button))
                }
            }
        }
    }
}

@Composable
private fun FailureBlock(message: String, onRetry: () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                imageVector = Icons.Filled.ErrorOutline,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.error,
                modifier = Modifier.size(20.dp),
            )
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = stringResourceSafe(R.string.profile_cert_failure_title),
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.error,
            )
        }
        Text(
            text = message,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.85f),
        )
        OutlinedButton(onClick = onRetry, shape = MaterialTheme.shapes.large) {
            Icon(imageVector = Icons.Filled.Refresh, contentDescription = null)
            Spacer(modifier = Modifier.width(8.dp))
            Text(text = stringResourceSafe(R.string.profile_cert_retry_button))
        }
    }
}

@Composable
private fun DetailRow(
    label: String,
    value: String,
    mono: Boolean = false,
    highlight: Color? = null,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.Top,
    ) {
        Text(
            text = label,
            modifier = Modifier.widthIn(min = 110.dp),
            style = MaterialTheme.typography.labelLarge,
            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.75f),
        )
        Spacer(modifier = Modifier.width(12.dp))
        Text(
            text = value,
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = if (highlight != null) FontWeight.SemiBold else FontWeight.Normal,
            fontFamily = if (mono) FontFamily.Monospace else null,
            color = highlight ?: MaterialTheme.colorScheme.onSurface,
        )
    }
}

@Composable
private fun InfoBanner(text: String, isError: Boolean, onDismiss: () -> Unit) {
    val container = if (isError) {
        MaterialTheme.colorScheme.errorContainer
    } else {
        MaterialTheme.colorScheme.primaryContainer
    }
    val content = if (isError) {
        MaterialTheme.colorScheme.onErrorContainer
    } else {
        MaterialTheme.colorScheme.onPrimaryContainer
    }
    Surface(
        modifier = Modifier.fillMaxWidth(),
        shape = MaterialTheme.shapes.medium,
        color = container,
        contentColor = content,
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = if (isError) Icons.Filled.ErrorOutline else Icons.Filled.CheckCircle,
                contentDescription = null,
                modifier = Modifier.size(18.dp),
            )
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = text,
                style = MaterialTheme.typography.bodySmall,
                modifier = Modifier.weight(1f),
            )
            TextButton(onClick = onDismiss, contentPadding = androidx.compose.foundation.layout.PaddingValues(horizontal = 4.dp)) {
                Text(text = stringResourceSafe(R.string.profile_cert_dismiss))
            }
        }
    }
}

@Composable
private fun stringResourceSafe(id: Int): String = androidx.compose.ui.res.stringResource(id)

File: iOS/Crisis Connect/Features/Profile/ProfileCertificateCard.swift
//
//  ProfileCertificateCard.swift
//  Crisis Connect
//
//  SwiftUI port of Android's `ProfileCertificateCard` (screens/settings).
//  Renders the device-bound rescue certificate. State and actions live on
//  `ProfileViewModel` so the profile header (role chip, verified badge) and the
//  rescue gating stay in sync with revoke / provision actions taken here.
//

import SwiftUI

struct ProfileCertificateCard: View {
    @ObservedObject var viewModel: ProfileViewModel
    @State private var confirmRevoke = false

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            header

            content

            if let statusMessage = viewModel.certificateState.statusMessage {
                infoBanner(text: statusMessage, isError: false)
            }
            if let errorMessage = viewModel.certificateState.errorMessage {
                infoBanner(text: errorMessage, isError: true)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .appSurface(style: .regular, padding: 18)
        .task { viewModel.refreshCertificate() }
        .alert(
            Text("PROFILE_CERT_REVOKE_TITLE"),
            isPresented: $confirmRevoke
        ) {
            Button("PROFILE_CERT_REVOKE_CONFIRM", role: .destructive) {
                viewModel.revokeCertificate(reason: "user_request")
            }
            Button("PROFILE_CERT_CANCEL", role: .cancel) {}
        } message: {
            Text("PROFILE_CERT_REVOKE_BODY")
        }
    }

    // MARK: - Header

    private var header: some View {
        HStack(alignment: .top, spacing: 12) {
            AppIconBadge(systemName: "checkmark.shield.fill", tint: .appPrimary)

            VStack(alignment: .leading, spacing: 4) {
                Text("PROFILE_CERT_CARD_TITLE")
                    .font(.headline.weight(.semibold))
                Text("PROFILE_CERT_CARD_SUBTITLE")
                    .font(.footnote)
                    .foregroundStyle(Color.appTextSecondary)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 8)

            if case let .loaded(_, isExpired, isRevoked) = viewModel.certificateState.status {
                statusChip(isExpired: isExpired, isRevoked: isRevoked)
            }
        }
    }

    private func statusChip(isExpired: Bool, isRevoked: Bool) -> some View {
        let label: LocalizedStringKey
        let tint: Color
        let symbol: String
        if isRevoked {
            label = "PROFILE_CERT_STATUS_REVOKED"; tint = .appDanger; symbol = "xmark.circle.fill"
        } else if isExpired {
            label = "PROFILE_CERT_STATUS_EXPIRED"; tint = .appWarning; symbol = "clock.fill"
        } else {
            label = "PROFILE_CERT_STATUS_ACTIVE"; tint = .appSuccess; symbol = "checkmark.circle.fill"
        }
        return HStack(spacing: 6) {
            Image(systemName: symbol)
                .font(.caption2.weight(.bold))
            Text(label)
                .font(.caption.weight(.semibold))
        }
        .foregroundStyle(tint)
        .padding(.horizontal, 10)
        .padding(.vertical, 5)
        .background(Capsule(style: .continuous).fill(tint.opacity(0.14)))
        .overlay(Capsule(style: .continuous).stroke(tint.opacity(0.24), lineWidth: 1))
        .fixedSize()
    }

    // MARK: - State content

    @ViewBuilder
    private var content: some View {
        switch viewModel.certificateState.status {
        case .loading:
            loadingBlock
        case .missing:
            missingBlock
        case let .loaded(certificate, isExpired, isRevoked):
            loadedBlock(certificate: certificate, isExpired: isExpired, isRevoked: isRevoked)
        case let .failure(message):
            failureBlock(message: message)
        }
    }

    private var loadingBlock: some View {
        HStack(spacing: 10) {
            ProgressView()
            Text("PROFILE_CERT_LOADING")
                .font(.subheadline)
                .foregroundStyle(Color.appTextSecondary)
        }
    }

    private var missingBlock: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("PROFILE_CERT_MISSING_BODY")
                .font(.subheadline)
                .foregroundStyle(.primary)
                .fixedSize(horizontal: false, vertical: true)

            Button {
                viewModel.provisionCertificate()
            } label: {
                busyLabel(
                    isBusy: viewModel.certificateState.provisioning,
                    busyTitle: "PROFILE_CERT_REQUEST_IN_PROGRESS",
                    title: "PROFILE_CERT_REQUEST_BUTTON",
                    systemImage: "lock.fill",
                    onPrimary: true
                )
            }
            .buttonStyle(AppPrimaryButtonStyle())
            .disabled(viewModel.certificateState.provisioning)
        }
    }

    private func loadedBlock(certificate: RoleCertificate, isExpired: Bool, isRevoked: Bool) -> some View {
        VStack(alignment: .leading, spacing: 14) {
            hero(certificate: certificate, isExpired: isExpired, isRevoked: isRevoked)

            VStack(spacing: 8) {
                detailRow(
                    label: "PROFILE_CERT_ROLE_LABEL",
                    value: certificate.role.prefix(1).uppercased() + certificate.role.dropFirst()
                )
                detailRow(label: "PROFILE_CERT_DEVICE_LABEL", value: certificate.deviceId, mono: true)
                detailRow(label: "PROFILE_CERT_ISSUED_LABEL", value: Self.formatDate(certificate.issuedAtMillis))
                detailRow(
                    label: "PROFILE_CERT_EXPIRES_LABEL",
                    value: Self.formatDate(certificate.expiresAtMillis),
                    highlight: expiryHighlight(certificate: certificate, isExpired: isExpired)
                )
            }

            Divider().overlay(Color.appBorder)

            Text("PROFILE_CERT_MANAGE_LABEL")
                .font(.caption.weight(.semibold))
                .foregroundStyle(Color.appTextSecondary)
                .textCase(.uppercase)

            manageActions(isExpired: isExpired, isRevoked: isRevoked)
        }
    }

    @ViewBuilder
    private func manageActions(isExpired: Bool, isRevoked: Bool) -> some View {
        let canRenew = isExpired || isRevoked
        HStack(spacing: 10) {
            if canRenew {
                Button {
                    viewModel.provisionCertificate()
                } label: {
                    busyLabel(
                        isBusy: viewModel.certificateState.provisioning,
                        busyTitle: "PROFILE_CERT_REQUEST_IN_PROGRESS",
                        title: "PROFILE_CERT_RENEW_BUTTON",
                        systemImage: "arrow.clockwise",
                        onPrimary: true
                    )
                }
                .buttonStyle(AppPrimaryButtonStyle())
                .disabled(viewModel.certificateState.provisioning)
            } else {
                Button {
                    viewModel.refreshCertificate()
                } label: {
                    Label("PROFILE_CERT_REFRESH_BUTTON", systemImage: "arrow.clockwise")
                }
                .buttonStyle(AppSecondaryButtonStyle())
            }

            Button {
                confirmRevoke = true
            } label: {
                busyLabel(
                    isBusy: viewModel.certificateState.revoking,
                    busyTitle: "PROFILE_CERT_REVOKE_BUTTON",
                    title: "PROFILE_CERT_REVOKE_BUTTON",
                    systemImage: "xmark.circle",
                    onPrimary: false,
                    tint: .appDanger
                )
            }
            .buttonStyle(AppSecondaryButtonStyle(tint: .appDanger))
            .disabled(viewModel.certificateState.revoking || isRevoked)
        }
    }

    /// Prominent success/warning banner so a freshly issued certificate is
    /// impossible to miss. Green = active (with the live "expires in …" time),
    /// amber = expired, red = revoked.
    private func hero(certificate: RoleCertificate, isExpired: Bool, isRevoked: Bool) -> some View {
        let tint: Color
        let symbol: String
        let title: LocalizedStringKey
        let subtitle: String
        if isRevoked {
            tint = .appDanger
            symbol = "xmark.seal.fill"
            title = "PROFILE_CERT_REVOKED_HEADLINE"
            subtitle = NSLocalizedString("PROFILE_CERT_INACTIVE_SUBTITLE", comment: "")
        } else if isExpired {
            tint = .appWarning
            symbol = "clock.fill"
            title = "PROFILE_CERT_EXPIRED_HEADLINE"
            subtitle = NSLocalizedString("PROFILE_CERT_INACTIVE_SUBTITLE", comment: "")
        } else {
            tint = .appSuccess
            symbol = "checkmark.seal.fill"
            title = "PROFILE_CERT_ACTIVE_HEADLINE"
            subtitle = String(
                format: NSLocalizedString("PROFILE_CERT_EXPIRES_RELATIVE", comment: ""),
                Self.relativeExpiry(certificate.expiresAtMillis)
            )
        }

        return HStack(spacing: 14) {
            Image(systemName: symbol)
                .font(.system(size: 30, weight: .semibold))
                .foregroundStyle(tint)
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.headline.weight(.bold))
                Text(subtitle)
                    .font(.subheadline)
            }
            Spacer(minLength: 0)
        }
        .foregroundStyle(tint)
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: AppTheme.cornerMedium, style: .continuous)
                .fill(tint.opacity(0.14))
        )
    }

    private func failureBlock(message: String) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 8) {
                Image(systemName: "exclamationmark.triangle.fill")
                    .foregroundStyle(Color.appDanger)
                Text("PROFILE_CERT_FAILURE_TITLE")
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(Color.appDanger)
            }
            Text(message)
                .font(.footnote)
                .foregroundStyle(Color.appTextSecondary)
                .fixedSize(horizontal: false, vertical: true)
            Button {
                viewModel.refreshCertificate()
            } label: {
                Label("PROFILE_CERT_RETRY_BUTTON", systemImage: "arrow.clockwise")
            }
            .buttonStyle(AppSecondaryButtonStyle())
        }
    }

    // MARK: - Rows & banners

    private func detailRow(
        label: LocalizedStringKey,
        value: String,
        mono: Bool = false,
        highlight: Color? = nil
    ) -> some View {
        HStack(alignment: .top, spacing: 12) {
            Text(label)
                .font(.subheadline.weight(.medium))
                .foregroundStyle(Color.appTextSecondary)
                .frame(minWidth: 96, alignment: .leading)
            Text(value)
                .font(mono ? .system(.subheadline, design: .monospaced) : .subheadline.weight(highlight != nil ? .semibold : .regular))
                .foregroundStyle(highlight ?? .primary)
                .frame(maxWidth: .infinity, alignment: .leading)
                .textSelection(.enabled)
        }
    }

    private func infoBanner(text: String, isError: Bool) -> some View {
        let tint: Color = isError ? .appDanger : .appSuccess
        return HStack(spacing: 8) {
            Image(systemName: isError ? "exclamationmark.circle.fill" : "checkmark.circle.fill")
                .font(.subheadline)
            Text(text)
                .font(.footnote)
                .frame(maxWidth: .infinity, alignment: .leading)
            Button("PROFILE_CERT_DISMISS") {
                viewModel.consumeCertificateMessages()
            }
            .font(.footnote.weight(.semibold))
            .buttonStyle(.plain)
        }
        .foregroundStyle(tint)
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .background(
            RoundedRectangle(cornerRadius: AppTheme.cornerMedium, style: .continuous)
                .fill(tint.opacity(0.12))
        )
    }

    // MARK: - Helpers

    @ViewBuilder
    private func busyLabel(
        isBusy: Bool,
        busyTitle: LocalizedStringKey,
        title: LocalizedStringKey,
        systemImage: String,
        onPrimary: Bool,
        tint: Color = .appPrimary
    ) -> some View {
        if isBusy {
            HStack(spacing: 8) {
                ProgressView()
                    .tint(onPrimary ? Color.white : tint)
                Text(busyTitle)
            }
        } else {
            Label(title, systemImage: systemImage)
        }
    }

    private func expiryHighlight(certificate: RoleCertificate, isExpired: Bool) -> Color? {
        if isExpired {
            return .appDanger
        }
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let dayMillis: Int64 = 24 * 60 * 60 * 1000
        if certificate.expiresAtMillis - now < dayMillis {
            return .appWarning
        }
        return nil
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter
    }()

    private static let relativeFormatter: RelativeDateTimeFormatter = {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .full
        return formatter
    }()

    private static func formatDate(_ millis: Int64) -> String {
        dateFormatter.string(from: Date(timeIntervalSince1970: Double(millis) / 1000))
    }

    private static func relativeExpiry(_ millis: Int64) -> String {
        relativeFormatter.localizedString(
            for: Date(timeIntervalSince1970: Double(millis) / 1000),
            relativeTo: Date()
        )
    }
}
